File: sdks/go/config_push.go
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/couchbase/gocb/v2"
	"github.com/go-faker/faker/v4"
)

var stdout = log.New(os.Stdout, ", ", log.Ldate|log.Lmicroseconds|log.Lmsgprefix)

func operationLogger(msg string) {
	stdout.Println(msg)
}

var params struct {
	host     string
	bucket   string
	username string
	password string

	keys               []string
	timeout            int
	configPollInterval int
	threads            int
	items              int
	runningTime        int
}

type Task struct {
	coll  *gocb.Collection
	index int
	mu    sync.Mutex
	stop  bool
}

func (t *Task) run(wg *sync.WaitGroup) {
	for !t.stop {
		key := t.nextKey()

		_, err := t.coll.Get(key, nil)
		if err != nil {
			operationLogger("[FAILURE] Read: " + err.Error())
			continue
		}

		_, err = t.coll.Upsert(key, getRandomDoc(), nil)

		if err != nil {
			operationLogger("[FAILURE] Store: " + err.Error())
		} else {
			operationLogger("[SUCCESS] Store: " + key)
		}
	}
	wg.Done()
}

func (t *Task) nextKey() string {
	t.mu.Lock()
	var key = params.keys[t.index]
	t.index = (t.index + 1) % params.items
	t.mu.Unlock()
	return key
}

func (t *Task) waiter(duration int) {
	time.Sleep(time.Duration(duration) * time.Second)
	t.stop = true
}

type SimpleData struct {
	Field1 string
	Field2 string
	Field3 string
	Field4 string
	Field5 string
}

func getRandomDoc() SimpleData {
	doc := SimpleData{}
	err := faker.FakeData(&doc)
	if err != nil {
		fmt.Println(err)
	}
	return doc
}

func argParse() {

	flag.StringVar(&params.host, "host", "localhost", "Cluster host address")
	flag.StringVar(&params.bucket, "bucket", "", "Bucket to use")
	flag.StringVar(&params.username, "username", "Administrator", "Username")
	flag.StringVar(&params.password, "password", "password", "Password")

	flag.IntVar(&params.timeout, "timeout", 2500, "Operation timeout")
	flag.IntVar(&params.configPollInterval, "c-poll-interval", 2500, "ConfigPollInterval")
	flag.IntVar(&params.threads, "num-threads", 1, "Number of threads")
	flag.IntVar(&params.items, "num-items", 0, "Number of items to interact with")
	flag.IntVar(&params.runningTime, "time", 5, "How long to run the program")

	var keys string
	flag.StringVar(&keys, "keys", "", "Doc keys to interact with")
	flag.Parse()

	params.keys = strings.Split(keys, ",")
	if params.items == 0 {
		params.items = len(params.keys)
	}
	fmt.Fprintf(os.Stderr, "Running benchmark for %d sec on host %s (%s) with [items: %d, threads: %d, timeout: %dus] \n",
		params.runningTime, params.host, params.bucket, params.items, params.threads, params.timeout)
}

func main() {
	argParse()

	gocb.SetLogger(gocb.VerboseStdioLogger())
	cluster, err := gocb.Connect("couchbase://"+params.host, gocb.ClusterOptions{
		Authenticator: gocb.PasswordAuthenticator{
			Username: params.username,
			Password: params.password,
		},
		TimeoutsConfig: gocb.TimeoutsConfig{
			KVTimeout: time.Duration(params.timeout) * time.Millisecond,
		},
	})
	if err != nil {
		log.Fatal(err)
	}

	bucket := cluster.Bucket(params.bucket)

	err = bucket.WaitUntilReady(5*time.Second, nil)
	if err != nil {
		log.Fatal(err)
	}
	task := Task{
		coll:  bucket.Scope("_default").Collection("_default"),
		stop:  false,
		index: 0,
	}

	var waitGroup sync.WaitGroup
	for i := 0; i < params.threads; i++ {
		go task.run(&waitGroup)
	}
	waitGroup.Add(params.threads)
	go task.waiter(params.runningTime)
	waitGroup.Wait()
}

File: sdks/python/config_push.py

import concurrent.futures
import logging
import sys
from argparse import ArgumentParser
from datetime import timedelta
from time import perf_counter_ns, sleep

from couchbase.auth import PasswordAuthenticator
from couchbase.cluster import Cluster
from couchbase.collection import CBCollection
from couchbase.options import ClusterOptions, ClusterTimeoutOptions, UpsertOptions


class ConfigPush:

    def __init__(self, coll: CBCollection, keys: list[str], threads: int, timeout: int):
        self.coll = coll
        self.threads = threads
        self.keys = keys
        self.done = False
        self.upsert_timeout = timeout

    def config_push_task(self, thread: int):
        key_index = 0
        doc = SimpleDocument()
        first_error = None
        next_success = None
        exception_msg = None
        while True:
            try:
                key = self.keys[(key_index := ((key_index + 1) % len(self.keys)))]
                self.coll.upsert(key, doc.next(key),
                                 UpsertOptions(timeout=timedelta(milliseconds=self.upsert_timeout)))
                if first_error:
                    next_success = perf_counter_ns()
                    break
            except Exception as e:
                if not first_error:
                    first_error = perf_counter_ns()
                    exception_msg = e

        time_taken = next_success - first_error
        logging.info('[FIRST FAILURE] {}'.format(first_error))
        logging.info('[NEXT SUCCESS] {}'.format(next_success))
        logging.info('[TIME TAKEN (s)] {}'.format(round(time_taken/1e9, 2)))
        logging.info('Exception: {}'.format(exception_msg))

    def run(self):
        self.config_push_task(0)

    def run_with_threading(self, runtime: int):
        with concurrent.futures.ThreadPoolExecutor(max_workers=self.threads) as executor:
            executor.map(self.config_push_task, range(self.threads))

            sleep(runtime)
            self.done = True


class SimpleDocument:

    @staticmethod
    def hex_digest(key: str) -> str:
        return '%032x' % key.__hash__()

    @staticmethod
    def build_email(alphabet: str) -> str:
        return '%s@%s.com' % (alphabet[12:18], alphabet[18:24])

    @staticmethod
    def build_string(alphabet: str):
        return '%s %s' % (alphabet[16:24], alphabet[24:])

    def next(self, key: str) -> dict:
        alphabet = self.hex_digest(key) + self.hex_digest(key[::-1])
        return {
            'name': self.build_string(alphabet),
            'email': self.build_email(alphabet),
            'count': ord(alphabet[-1])
        }


def get_args():
    parser = ArgumentParser()

    parser.add_argument('--host', default='127.0.0.1',
                        help='the host string/ip')
    parser.add_argument('--bucket', required=True,
                        help='bucket name')
    parser.add_argument('--username', default='Administrator',
                        help='cluster username')
    parser.add_argument('--password', default='password',
                        help='password')
    parser.add_argument('--num-threads',
                        dest='threads',
                        default=1,
                        help='threads to use')
    parser.add_argument('--timeout',
                        default=2500,
                        help='kv timeout (ms)')
    parser.add_argument('--keys', required=True,
                        help='keys to process')
    parser.add_argument('--time',
                        default=10,
                        help='the time to run the experiment (s)')
    parser.add_argument('--c-poll-interval',
                        dest='cp_interval',
                        default=2500,
                        help='ms')
    parser.add_argument('extras',
                        nargs='*',
                        help='extra settings will be ignored')
    return parser.parse_known_args()


def main():
    args, _ = get_args()
    logging.basicConfig(level=logging.INFO, stream=sys.stdout,
                        format='%(asctime)s.%(msecs)03d, %(message)s',
                        datefmt='%Y-%m-%d %H:%M:%S')
    c_options = ClusterOptions(PasswordAuthenticator(args.username, args.password),
                               timeout_options=ClusterTimeoutOptions(
                                   kv_timeout=timedelta(milliseconds=int(args.timeout))),
                               config_poll_interval=timedelta(milliseconds=int(args.cp_interval)))
    cluster = Cluster('couchbase://{}'.format(args.host), c_options)
    cluster.wait_until_ready(timedelta(seconds=5))

    cb_coll = cluster.bucket(args.bucket).default_collection()
    threads = int(args.threads)
    test = ConfigPush(cb_coll, args.keys.split(','), threads, int(args.timeout))
    if threads > 1:
        test.run_with_threading()
    else:
        test.run()


if __name__ == '__main__':
    main()
